staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go

```diff
@@ -600,18 +600,24 @@ func (q *Quantity) Neg() {
 const int64QuantityExpectedBytes = 18
 
 // String formats the Quantity as a string, caching the result if not calculated.
-// String is an expensive operation and caching this result significantly reduces the cost of
-// normal parse / marshal operations on Quantity.
 func (q *Quantity) String() string {
```

Comment:
`resource.Quantity` seems should be an immutable object. Current logic of `String()` obviously  doesn't respect any change on the object. Why r u trying to modify it at the first place?

Reply:
`resource.Quantity` looks like a mutable object to me... take a look at some of it's functions, values of the properties attached to the entity change over time. Mutating a `resource.Quantity` can cause problems when that `resource.Quantity` is sourced from the shared cache (as part of some other resource).

Let's say resource `x` uses this entity and has `Quantity` fields associated with it (i.e. `x.expectedSize`). If thread **A** and **B** attempt to call `x.expectedSize.String()` around the time `t0`, then there would exist a potential race condition where they both end up [mutating](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L610) the internal field `s` of `x.expectedSize`. From what I see it's not wrapped around a mutex or a thread-safe var. Some of the other functions also seem to [clear](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L525) the string field `s`. To expand on this example, if thread **C** is clearing the `x.expectedSize` field at time `t1`, then thread A and B will be returning an empty string value when accessing [this line](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L612) at `t1`.

If you want a more practical reason for this change; we have logging in place that could result in `x.expectedSize.String()` being called concurrently.

I do understand that quantity isn't intended to be thread safe, and we're working around that on our end to somewhat make it so. That being said, this change would be of great convenience to us, and hopefully others too. Right now, we clone the quantity and string that to not modify the original, which isn't quite ideal.